### DISCUSSION TITLE
alternative `zustand/context` implementation

### DISCUSSION THREAD

**Discussion by lubieowoce** [pmndrs/zustand#566] [Ideas]
Per my comment in [Can't auto generate selectors when using Context? #549](https://github.com/pmndrs/zustand/issues/549#issuecomment-922470447).
For background, i'm interested in using zustand for component-local state, so zustand/context is a necessity.

The improvement here would be a slightly more normal implementation of `zustand/context` - instead of [passing a hook function down via context](https://github.com/pmndrs/zustand/blob/main/src/context.ts#L60), the context would only provide the stuff that's needed, i.e. the API object. AFAICT this would also cut down the size of `zustand/context`, because you can avoid all the wrapping it does.
The (possibly naive) implementation idea was to basically just add a layer of indirection in `useStore` (`getApi`, see below) to be able to get the `api` object from `useContext()`, something like this:

``` tsx
// factored out of `create()`
// intended usage:
//    createUseStoreHook(() => api)
//    createUseStoreHook(() => useContext(SomeContext))

const createUseStoreHook = <TState,>(getApi: () => StoreApi<TState>) => {
  const useStore = (selector, equalityFn) => {
    const api = getApi();
    // ... the usual stuff
  }
  return useStore
}


export const create = (createState) => {
  const api = ...
  return createUseStoreHook(() => api)
}


export const createContext = (createState) => {
  const ZustandContext = reactCreateContext()
  const useStore = createUseStoreHook(() => useContext(ZustandContext));
  const Provider = ({ children }) => {
    const apiRef = useRef();
    if (!apiRef.current) {
      apiRef.current = create(createState)
    }
    return <ZustandContext.Provider value={apiRef.current}>{children}<ZustandContext.Provider>
  }
  return { Provider, useStore }
}
```
 
Also re: https://github.com/pmndrs/zustand/issues/549#issuecomment-922472178 @dai-shi you mentioned that the with a `useSyncExternalStore`-based implementation, `zustand/context` could be removed altogether. i'd love to hear your vision for how that would work! wouldn't you need a context to pass the correct `subscribe` function to `uSES` anyway?

**Comment by dai-shi** [comment]
Yes, I now see my idea is basically the same as yours. The API design is slightly different, but the approach is the same.

All we need is to expose a new `useStore(store, selector, equalityFn)` API. Let's assume we export it from `zustand/context`. It will be like this.

```jsx
import { createContext, useContext } form 'react'
import create from 'zustand/vanilla'
import { useStore } from 'zustand/context'
import useConstant from 'use-constant';

const ZustandContext = createContext()
const createStore = () => create(...)

export const ZustandProvider = ({ children }) => {
  const store = useConstant(() => createStore())
  return (
    <ZustandContext.Provider value={store}>
      {children}
    </ZustandContext.Provider>
  )
}

export const useZustandStore = (selector, equalityFn) => {
  const store = useContext(ZustandContext)
  return useStore(store, selector, equalityFn)
}
```

I had this idea from v3, but at that point we were going to use `uMS` in v4. And, this wasn't possible. (`zustand/vanilla` can't depend on react.)
But as now we see, with `uSES`, this will be possible. After investigation, I will probably include it in v4.0 or v4.1.

**Reply by dai-shi** [reply]
Done in 4.0.0-alpha.3. (Not finalized.)

**Reply by lubieowoce** [reply]
oh hey, i like yours better :D making the API object a param feels obvious in retrospect, it's way simpler this way!

**Comment by lubieowoce** [comment]
Which got me thinking: what would happen if the user swapped contexts at runtime somehow, effectively changing which API object gets passed to `useStore`? I assume nothing good, although maybe it'd work out fine?

**Reply by dai-shi** [reply]
It should work. You mean like this?

```jsx
import { createContext, useContext } form 'react'
import create from 'zustand/vanilla'
import { useStore } from 'zustand' // 4.0.0-alpha.3

const ZustandContext = createContext()

const store1 = create(...)
const store2 = create(...)

export const ZustandProvider = ({ something, children }) => {
  return (
    <ZustandContext.Provider value={something ? store1 : store2}>
      {children}
    </ZustandContext.Provider>
  )
}

export const useZustandStore = (selector, equalityFn) => {
  const store = useContext(ZustandContext)
  return useStore(store, selector, equalityFn)
}
```

If it doesn't, we will investigate (because it's the whole point of this change.)

**Reply by AndrewJSchoen** [reply]
@dai-shi I am attempting to do this now, but with Typescript and, for good measure, the new `useStoreWithEqualityFn` from `zustand/traditional`. 

```typescript
import { createContext, useContext } from "react";
import { useStoreWithEqualityFn } from "zustand/traditional";
import { MyState, MyStore } from "./types";
// MyStore is ReturnType<typeof RETURN-FROM-CREATE-FN>

const MyContext = createContext<MyStore | null>(null);

export function useMyStore<MyState>(
  selector: (state: MyState) => MyState,
  equalityFn?: (left: MyState, right: MyState) => boolean
) {
  const store = useContext(MyContext);
  if (!store)
    throw new Error("Missing MyContext.Provider in the tree");
  return useStoreWithEqualityFn(store, selector, equalityFn);
}
```

But I get an error on the `store` parameter of the `useStoreWithEqualityFn`:
```
Argument of type '{}' is not assignable to parameter of type 'WithReact<StoreApi<unknown>>'.
  Type '{}' is missing the following properties from type 'StoreApi<unknown>': setState, getState, subscribe, destroy ts(2345)
```

The provider was essentially constructed the same way as above, but gives some odd errors. Not sure if it is just a consequence of the above, though:

```typescript
export function MyProvider({
  store,
  children,
}: {
  store: MyStore | undefined;
  children: ReactNode;
}): ReactNode {
  return (
    <MyContext.Provider value={store ? store : createMyStore}>
      {children}
    </MyContext.Provider>
  );
};
```
Error: 
```
Operator '>' cannot be applied to types '{ store: UseBoundStore<Write<StoreApi<MyState>, StoreSubscribeWithSelector<MyState>>>; }' and '{ children: ReactNode; }'.ts(2365)
Operator '<' cannot be applied to types 'boolean' and 'RegExp'.ts(2365)
```

For what it's worth, I did try this with and without the equality functions, and didn't really get a difference. Is this the result of changes to the API, or am I specifying something incorrectly?

**Reply by dai-shi** [reply]
Can you complete the reproduction in typescript playground? https://tsplay.dev/N7VARw

**Reply by AndrewJSchoen** [reply]
Hi @dai-shi,

Here is the reproduction. I added both standard and equality function versions (just comment/uncomment the relevant sections):
[link here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQYQgO2wA8YAaOAVwGdcDj4BfOAMyghDgCJ0NkZOA3AChQkWIhQ9s5amgDKMaGjhNW7OAHIAXtRgZ8AEw3DR0eElkKlAdWAwAFgFEAjhQwAbOwE8AYvnKomNi2Di5unjC++CosbBycOlR6hgD0MFAYBnbABB6CIuBmElT2Hu4QAO4xavGJyQYpJWWV+abiSFQUAEZUqMBdaCH2cmjuaHzQ1XFcdfoNIMAGBmMVGOj5KSlwALJeVuhwwFRwAEpoMBRQ+AAqXmBoADyR9xDMp47XAKonAHIAtD4TgB5bZ-HAnRwAQWujgBPwAfEIRIQ0FBmLxlLsFFhlAghHA4AM1lQAFxwfAUEADKDCAnAfDIMkACgpIDJrOpAEo4ABeeFwABuEEWQgYSOQBCScAMaHRFHcMH2yh5cCZNDIAH0AObnbl81V4glEqCkuAABlI+MODOZrPZlK5vP56qZTOQl3QhDJWL02D1-KZCGNpPdUE9MAAdMG4ABqcmUhiczmi5NCCX4KUAEVlGHlis8yGVcE6PT6AyGIzGEygDx9OPhTJlcoVSs5wiEmzgYQ83jgPgoDJgOWiADVUVRh2nJfBAjilbzJEFBnYnK4e5E-LW9r60A3OUzs8388BC6Qmu5yhU20jO9jDGsDHAxybJ5301LZ9h5yrP4869g9wPHM8zkAs0FTZ5MW3JQFzOC4rlue4njuNBXkXOdFHQeF23feBdjwFESAXX8CPoLd5wAH3jC8GwpC9rw7LZuwiLw+wHPhhyfcdJzQIgxHgZh2KHAhKBoH0lHIncGytGgq0w5kkhxb1txxf0dhU6QrTQNcWL8AB+ZkxmYGBlOxaQ4CgYAtXsEz1LM8CnUJCAIDGfQhG5Q0UGnYtMKLWRSN4mAmXwuhAuvOk3iZABCJIlGTAkCQcNgqnwNAqkcMNoCZThtiOCd8C1dSApICMAAU2AFRZUWtOAHGUdI0DQThwos85LmiSxfKGZjvD8NVfPIWTxkw8htPCXr8GvMVGLgO8DAfLiXwIGbeP4lghM42RxPQST6yZGaCSG6sFJ3UydzU-80EtTsCTG9cogM1UjNsy7yEs6yXo0hz9S6ZzXPwGaPIOryM3gWKDhVfzQpIYKvGKmBrxuw5Ipi3zkyRxL7GS8k0q7TKoGy3KqHywqQsIyNyogSqZSgGq6tq9AmsRrYCXQeCOpoJV+qUQbRmG6BmZUIA)

Let me know if you have any questions.

**Reply by dai-shi** [reply]
You should use `createStore` for the context usage.

Here's the fixed one: https://tsplay.dev/NVOPMm

**Reply by AndrewJSchoen** [reply]
Ah, that fixes it. Thanks!
